### PR TITLE
Prototype Asyncio-based Crawler

### PR DESCRIPTION
**Pull request by wdm0006**
In particular, note the class in crawlers/base.py.  It is a self contained asyncio-powered crawler, that scales on one machine pretty nicely, since so much of this task is IO bound.  For a bit of an idea of performance, pulling down a bunch of google search pages with the DefaultEngine with different numbers of workers takes:
```
workers=1, N-pages=100, 100.97s
workers=3, N-pages=100, 18.1s
workers=10, N-pages=100, 9.3s
```
It does currently make an assumption of a bounded input, so you pass in the urls collection and it will churn through that and then be done.  That collection can be a generator though, so we can probably be creative there for large lists of URLs.  

There is a use-case where when scraping one URL, you find another that you'd also like to scrape, that isn't really neatly supported here (maybe the consumers can directly put into the queue?).  Also what to actually do with the output object isn't really fleshed out (maybe pass a callback into the `__init__`?).

 Looking for some feedback.  My thinking was that a decent UX for the crawlers would be something like:

```
data = (foo for foo in bar) # some collection of urls
engine = DefaultEngine() # define some engine

# pass engine, data and kwargs to a crawler of some sort
crawler = AsyncBoundedCrawler(engine, data, **kwargs)

# blocking call to run() to process the urls.
crawler.run()
```

Is that more of less what ya'll had in mind?

### CODE REVIEW

File: rasp/crawlers/asyncio_crawler.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            url = await self._queue.get()
+            if url is None:
+                self._queue.task_done()
+                break
+            else:
+                wp = engine.get_page_source(url)
+                self._callback(wp)
+                self._queue.task_done()
+
+    def run(self):
```
Comment (dlrobertson) [inline rasp/crawlers/asyncio_crawler.py:50]: I feel like we should have a non-blocking alternative. E.g. `run(blocking=False)` and a `is_finished` condition variable. I'm a fan of incremental development, so I'd suggest punting on this feature for now and opening an issue.

File: rasp/crawlers/asyncio_crawler.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    async def consumer(self, consumer_id):
+        engine = copy.deepcopy(self._engine)
+        while not self._queue.empty():
+            url = await self._queue.get()
+            if url is None:
+                self._queue.task_done()
+                break
+            else:
+                wp = engine.get_page_source(url)
+                self._callback(wp)
```
Comment (dlrobertson) [inline rasp/crawlers/asyncio_crawler.py:47]: Definitely agree that in the future we should add the capability for someone to implement a type of spider crawler with this.

File: rasp/crawlers/asyncio_crawler.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    def run(self):
+        try:
+            self._loop.run_until_complete(self._main())
+        finally:
+            for task in asyncio.Task.all_tasks():
+                task.cancel()
+            self._loop.close()
+
+    def stop(self):
+        self._loop.close()
```
Comment (dlrobertson) [inline rasp/crawlers/asyncio_crawler.py:59]: This will leak processes when using the `SeleniumEngine` right?

**Review by dlrobertson — COMMENTED** [review]
Just did my first pass. Looks really good! I'll want to make a second pass at some point.

**Comment by dlrobertson** [comment]
> There is a use-case where when scraping one URL, you find another that you'd also like to scrape, that isn't really neatly supported here (maybe the consumers can directly put into the queue?).

I'm always wary of giving consumers direct access to class members. We certainly could provide a handle to the queue. (History is also very important, but I lean towards letting the user take care of that). I'm also wary of providing a raw handle to the queue. If we're feeling cautious we could provide a handle to a list that we check after we call the callback and en-queue any existing urls? just a thought.

**Comment by dlrobertson** [comment]
Navigating conflicts between #23 and this will also be difficult.

**Comment by wdm0006** [comment]
Re: #23, this isn't intended to be merged any time soon, and the two implemented have no parallelism, just concurrency within a single thread, so the functional representation wouldn't be needed.  For true parallelism and more distributed schemes, it will be more useful.  Once merged, I don't see a reason not to just standardize on using the curried version internally in crawlers, for consistency.   I'll handle the merge conflicts once we've got an idea of what we actually want to do with the crawlers.

As for restricting access to the queue for users, yeah, it's certainly possible for people to make endless loops for themselves, or get themselves in trouble otherwise.  Strictly speaking, they have access either way for good or for ill in python, so my tendency is to make it safe rather than try to restrict it whenever possible.   

That said, I don't have a good idea of how to handle the case where you are getting more links to scrape as you go.  Writing them out somewhere, and then just repeating the process with the new list is certainly simplest, but requires some more work from the user.

File: rasp/crawlers/asyncio_crawler.py (hunk 2 above)
Comment (wdm0006) [reply rasp/crawlers/asyncio_crawler.py:47]: Yeah, that will take a bit of extra thought.  For now, the callback would be where something like palantiri would be able to parse a webpage object and write it out to psql or mongo.

File: rasp/crawlers/asyncio_crawler.py (hunk 1 above)
Comment (wdm0006) [reply rasp/crawlers/asyncio_crawler.py:50]: Agreed

File: rasp/crawlers/asyncio_crawler.py (hunk 3 above)
Comment (wdm0006) [reply rasp/crawlers/asyncio_crawler.py:59]: The consumers are long-ish lived (multiple requests each), so any cleanup required should be handled when the consumer closes.  I forgot to add in any calls to cleanup (or to use the context manager), but will add those in.

**Comment by hank-bond** [comment]
If/when we use the crawler in "production" we should probably have a hook to check if any given URL as already been scraped.  The viability of what has been discussed in this PR probably depends on how we see the user using the anidata suite.

If the crawler is running in batches it could probably just pull a list of hashed URLs from the storage database, but AFAIK the data only gets stored by the ETL?

Would some intermediary store for URLs be useful (and not overbearing)?  Maybe something like `redis`?  Maybe just a serialized list that gets deserialized and loaded in memory on startup of the crawler?

**Comment by wdm0006** [comment]
I'd say state like a list of those processed is out of scope of the crawler itself, and would be handled in the callback.

**Comment by dlrobertson** [comment]
> If/when we use the crawler in "production" we should probably have a hook to check if any given URL as already been scraped. The viability of what has been discussed in this PR probably depends on how we see the user using the anidata suite.

Very true. Most crawlers need some sort of "History", but IMO this is outside the scope of this project. for a two reasons 1) Memory. It is super easy to get a history list too big for Python. 2) This would have to be more like a cache than a true list. The data in the history becomes stale at some point. A good number of very long running crawlers would want to revisit stale sites while others wouldn't.

> Would some intermediary store for URLs be useful (and not overbearing)? Maybe something like redis? Maybe just a serialized list that gets deserialized and loaded in memory on startup of the crawler?

Fantastic idea. You're definitely on to something here, but I think this should be implemented in the users code. In the users implementation both points above are still valid/concerns.

> I'd say state like a list of those processed is out of scope of the crawler itself, and would be handled in the callback.

Agreed.

**Comment by wdm0006** [comment]
so what do yall think about the general interface, and my assumption that we have 2 classes of crawler:

 * bounded: a fixed collection of urls to iterate through, then stop
 * unbounded: a runtime where the urls are considered an infinite stream (may require some external queuing mechanism).

So far I've only really touched the bounded case here, where I'm proposing we have some interface like:

```
data = (foo for foo in bar) # some collection of urls
engine = DefaultEngine() # define some engine

# pass engine, data and kwargs to a crawler of some sort
crawler = AsyncBoundedCrawler(engine, data, **kwargs)

# blocking call to run() to process the urls.
crawler.run()
```

Unbounded may be more involved, if there is an external queue reference somewhere instead of passing in data to the `__init__`. 

If the interface for the bounded case is agreeable, and ya'll agree with my slight restructuring of the repo, I can clean up the merge conflicts and we can iterate on the actual crawlers in smaller issues.  For the unbounded case, I think we probably need to have a bit more discussion, perhaps seperately.

**Comment by wdm0006** [comment]
My branch is all screwed up, I'll probably make a fresh one this week with a cleaned up version of the DefaultBoundedCrawler, and resubmit the AsyncioBoundedCrawler as a separate PR

**Pull request closed without merging**